### ISSUE TITLE
build regression caused by commit 48feaceb53fa

### ISSUE THREAD

**Issue by lersek** [openssl/openssl#8904]
In commit 48feaceb53fa ("Remove the possibility to disable the UI module entirely", 2017-07-03), first released in `OpenSSL_1_1_1`, the following preprocessing directives were removed:

    -#ifndef OPENSSL_NO_UI
    -#endif /* OPENSSL_NO_UI */

from file `crypto/evp/evp_key.c`. As a result, the `BUFSIZ` macro references in the code are no longer hidden, even if OpenSSL is configured with both `no-ui` and `no-stdio`. This causes a build breakage, given that the `BUFSIZ` macro would normally come from stdio.

The code should be fixed up in one of two ways:
* either reinstate the conditional directives, but reference `OPENSSL_NO_STDIO`,
* or locally define the `BUFSIZ` macro if it is not yet defined; similarly to file `crypto/ui/ui_util.c` (where this logic dates back to commit 984d6c605216, "Fix no-stdio build", 2015-09-29).

Thanks!

**Comment by levitte** [comment]
Have you noticed that `crypto/evp/evp_key.c` includes `<stdio.h>` unconditionally?

Either way, locally defining `BUFSIZ` if it's not already defined would be the way to go.

**Comment by lersek** [comment]
@levitte there is a sizeable set of `*.c` files that include `<stdio.h>` unconditionally, while at the same time containing code that's conditional on `OPENSSL_NO_STDIO`. At commit c5f7a99645aa:

    crypto/asn1/a_d2i_fp.c
    crypto/asn1/a_i2d_fp.c
    crypto/asn1/a_strex.c
    crypto/bio/b_dump.c
    crypto/bio/bio_cb.c
    crypto/bn/bn_print.c
    crypto/conf/conf_lib.c
    crypto/dh/dh_prn.c
    crypto/dsa/dsa_prn.c
    crypto/ec/eck_prn.c
    crypto/engine/eng_openssl.c
    crypto/err/err_prn.c
    crypto/lhash/lh_stats.c
    crypto/mem_dbg.c
    crypto/pem/pem_all.c
    crypto/pem/pem_info.c
    crypto/pem/pem_lib.c
    crypto/pem/pem_pk8.c
    crypto/pem/pem_pkey.c
    crypto/pkcs12/p12_utl.c
    crypto/rsa/rsa_prn.c
    crypto/x509/t_crl.c
    crypto/x509/t_req.c
    crypto/x509/t_x509.c
    crypto/x509/x_all.c
    crypto/x509v3/v3_prn.c
    ssl/ssl_txt.c

Anyway, what would be the recommended `BUFSIZ` value? Thank you.

**Comment by richsalz** [comment]
@lersek I think you're confused.  Richard pointed out that stdio.h is always included for the file you mention. Do you actually see a build failing, or is this theoretical?  If you see it fail, please post your config command and the compiler error output.

**Comment by richsalz** [comment]
@lersek I think you're confused.  Richard pointed out that stdio.h is always included for the file you mention. Do you actually see a build failing, or is this theoretical?  If you see it fail, please post your config command and the compiler error output.

**Comment by richsalz** [comment]
Or are you saying that with no-stdio you don't want to see stdio.h included at all?  That's a separate issue, and an enhancement request really.

**Comment by lersek** [comment]
Yes, this is an actual build failure.

In the [edk2 project](https://github.com/tianocore/edk2/), we're currently advancing our OpenSSL git submodule from `OpenSSL_1_1_0j` to `OpenSSL_1_1_1b`. I'm not myself implementing the git superproject (= edk2) changes for that, just reviewing the patches. Please refer to:
* https://bugzilla.tianocore.org/show_bug.cgi?id=1089
* https://www.redhat.com/archives/edk2-devel-archive/2019-May/msg00477.html

In the proposed patches, I noticed that edk2's `CryptoPkg/Library/Include/CrtLibSupport.h` file would now itself `#define BUFSIZ`, with a reference to OpenSSL's `evp_key.c`. This edk2 change looked wrong, given that `evp_key.c` has contained `BUFSIZ` references forever (specifically since commit a63d5eaab28a, from 2001). Therefore I tracked down the actual OpenSSL change that led to the submodule build failure, requiring the git superproject to provide `BUFSIZ`.

I cannot post the compiler output because I haven't (yet) reproduced the build issue myself -- again I'm just a reviewer for the edk2 patchset, and the submitter didn't attempt to track down the OpenSSL change this closely (and they didn't include the build error either). I can paste the `Configure` invocation however:

    # Configure UEFI
    system(
        "./Configure",
        "UEFI",
        "no-afalgeng",
        "no-asm",
        "no-async",
        "no-autoalginit",
        "no-autoerrinit",
        "no-bf",
        "no-blake2",
        "no-camellia",
        "no-capieng",
        "no-cast",
        "no-chacha",
        "no-cms",
        "no-ct",
        "no-deprecated",
        "no-dgram",
        "no-dsa",
        "no-dynamic-engine",
        "no-ec",
        "no-ec2m",
        "no-engine",
        "no-err",
        "no-filenames",
        "no-gost",
        "no-hw",
        "no-idea",
        "no-mdc2",
        "no-pic",
        "no-ocb",
        "no-poly1305",
        "no-posix-io",
        "no-rc2",
        "no-rfc3779",
        "no-rmd160",
        "no-scrypt",
        "no-seed",
        "no-sock",
        "no-srp",
        "no-ssl",
        "no-stdio",
        "no-threads",
        "no-ts",
        "no-ui",
        "no-whirlpool",
        # OpenSSL1_1_1b doesn't support default rand-seed-os for UEFI
        # UEFI only support --with-rand-seed=none
        "--with-rand-seed=none"
        ) == 0 ||
            die "OpenSSL Configure failed!\n";

**Comment by richsalz** [comment]
Someone needs to reproduce this or show the compiler output.  Oh well.

**Comment by lersek** [comment]
@richsalz I'm not proposing to make the inclusion of `<stdio.h>` conditional on `!defined(OPENSSL_NO_STDIO)`. The idea was to make only the `BUFSIZ` references (= the containing function definitions) conditional. Edk2 is firmware; it's not a hosted but a freestanding C environment, and its `stdio.h` doesn't provide `BUFSIZ`. This was the same when commit 984d6c6 was made, in 2015.

However, I agree it's perfectly fine if we conditionally define `BUFSIZ` on the spot (if it's not defined yet, despite having included `<stdio.h>`), and we ignore `OPENSSL_NO_STDIO` totally.

**Comment by lersek** [comment]
@richsalz I'm willing to reproduce the output, I'll just need a bit of time.

**Comment by richsalz** [comment]
Thanks for the details; I understand now.  I suggest something like add ```-DBUFSIZ=1024``` to your config line, maybe.

**Comment by lersek** [comment]
@richsalz that could work, yes -- but wouldn't it be more consistent to extend what commit 984d6c6 did, to the OpenSSL source files that have now become affected?

    git show 984d6c6 -- crypto/des/read2pwd.c crypto/ui/ui_util.c

From these, `crypto/des/read2pwd.c` exists no longer, but the pattern continues to exist in `crypto/ui/ui_util.c`. That's what I'd suggest for `crypto/evp/evp_key.c` as well. Thank you!

**Comment by richsalz** [comment]
Well, that platform isn't supported by openssl so waiting for it to be fixed might take awhile.  Adding the config param will make things work now.

**Comment by lersek** [comment]
@richsalz I'm perfectly willing to submit a pull request, once you guide me to the right approach :) We can add the workaround to edk2 now, but I'd like that to be temporary, and back it out once this OpenSSL issue report is closed / fixed. Thanks.

**Comment by lersek** [comment]
OK, so please see the `Configure` invocation in https://github.com/openssl/openssl/issues/8904#issuecomment-491019423 above. Now I can paste the error message too:

    "gcc" \
      -g \
      -fshort-wchar \
      -fno-builtin \
      -fno-strict-aliasing \
      -Wall \
      -Werror \
      -Wno-array-bounds \
      -ffunction-sections \
      -fdata-sections \
      -include AutoGen.h \
      -fno-common \
      -DSTRING_ARRAY_NAME=OpensslLibStrings \
      -m64 \
      -fno-stack-protector "-DEFIAPI=__attribute__((ms_abi))" \
      -maccumulate-outgoing-args \
      -mno-red-zone \
      -Wno-address \
      -mcmodel=small \
      -fpie \
      -fno-asynchronous-unwind-tables \
      -Wno-address \
      -O0 \
      -U_WIN32 \
      -U_WIN64 \
      -DL_ENDIAN \
      -DOPENSSL_SMALL_FOOTPRINT \
      -D_CRT_SECURE_NO_DEPRECATE \
      -D_CRT_NONSTDC_NO_DEPRECATE \
      -Wno-error=maybe-uninitialized \
      -Wno-error=format \
      -Wno-format \
      -Wno-error=unused-but-set-variable \
      -DNO_MSABI_VA_FUNCS \
      -mno-mmx \
      -mno-sse \
      -D DISABLE_NEW_DEPRECATED_INTERFACES \
      -c \
      -o $WORKSPACE/Build/Ovmf3264/NOOPT_GCC48/X64/CryptoPkg/Library/OpensslLib/OpensslLib/OUTPUT/openssl/crypto/evp/evp_key.obj \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/ssl/statem \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/ssl/record \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/ssl \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/x509v3 \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/x509 \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/ui \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/txt_db \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/stack \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/sm4 \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/sm3 \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/siphash \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/sha \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/rsa \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/rc4 \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/rand \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/pkcs7 \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/pkcs12 \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/pem \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/ocsp \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/objects \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/modes \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/md5 \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/md4 \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/lhash \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/kdf \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/hmac \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/evp \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/err \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/dso \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/dh \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/des \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/conf \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/comp \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/cmac \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/buffer \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/bn \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/bio \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/async \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/async/arch \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/asn1 \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/aria \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/aes \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib \
      -I$WORKSPACE/Build/Ovmf3264/NOOPT_GCC48/X64/CryptoPkg/Library/OpensslLib/OpensslLib/DEBUG \
      -I$WORKSPACE/MdePkg \
      -I$WORKSPACE/MdePkg/Include \
      -I$WORKSPACE/MdePkg/Include/X64 \
      -I$WORKSPACE/CryptoPkg \
      -I$WORKSPACE/CryptoPkg/Include \
      -I$WORKSPACE/CryptoPkg/Library/Include \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/include \
      -I$WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/include \
      $WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/evp/evp_key.c

    $WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/evp/evp_key.c: In function 'EVP_read_pw_string_min':
    $WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/evp/evp_key.c:52:15: error: 'BUFSIZ' undeclared (first use in this function)
         char buff[BUFSIZ];
                   ^
    $WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/evp/evp_key.c:52:15: note: each undeclared identifier is reported only once for each function it appears in
    $WORKSPACE/CryptoPkg/Library/OpensslLib/openssl/crypto/evp/evp_key.c:52:10: error: unused variable 'buff' [-Werror=unused-variable]
         char buff[BUFSIZ];
              ^

**Comment by richsalz** [comment]
The stdio header file on your target platform, one which is not an official platform supported by the project, doesn't define BUFSIZ (unlike almost all other platforms).  I think the best solution is for you to add the define, with a value that is appropriate for your platform and close this.  Perhaps others on the project disagree.

**Comment by lersek** [comment]
The following patch allows the build to complete:

    diff --git a/crypto/evp/evp_key.c b/crypto/evp/evp_key.c
    index e5ac107c385c..b8194b34cf34 100644
    --- a/crypto/evp/evp_key.c
    +++ b/crypto/evp/evp_key.c
    @@ -14,6 +14,10 @@
     #include <openssl/evp.h>
     #include <openssl/ui.h>

    +#ifndef BUFSIZ
    +#define BUFSIZ 256
    +#endif
    +
     /* should be init to zeros. */
     static char prompt_string[80];

**Comment by lersek** [comment]
> The stdio header file on your target platform, one which is not an official platform supported by the project, doesn't define BUFSIZ (unlike almost all other platforms).

Yes, that's indeed the case.

> I think the best solution is for you to add the define, with a value that is appropriate for your platform and close this.

I think this will introduce an inconsistency with currently existing code, in `crypto/ui/ui_util.c`, from commit 984d6c6. Then there will be two instances of the "no `BUFSIZ` from `<stdio.h>`" problem in the OpenSSL tree, and one of them will be worked around in-tree, while the other will be worked around out-of-tree.

Anyway: I wouldn't like to waste your time with this. Please re-confirm that you prefer us to add the `#define` to edk2, and I'll close this. Thanks for your attention!

**Comment by levitte** [comment]
We did add UEFI support not long ago (only a couple of years)... So it seems I introduced a bug, and I think it's *perfectly* appropriate to fix it, especially when it's that easy and the community is willing to provide.

**Comment by lersek** [comment]
@levitte thank you.

I assume the development workflow is for me to push a topic branch to my openssl clone on github (to be created), and to submit a new GitHub-based pull request for that. Is that correct?

Should I reference this ticket in the GitHub-based pull request? (AIUI such pull requests create GitHub tickets automatically, so ultimately we'd end up with two tickets.) Thanks!

**Comment by mspncp** [comment]
> I assume the development workflow is for me to push a topic branch to my openssl clone on github (to be created), and to submit a new GitHub-based pull request for that. Is that correct?

Yes, that's correct.

> Should I reference this ticket in the GitHub-based pull request? (AIUI such pull requests create GitHub tickets automatically, so ultimately we'd end up with two tickets.) Thanks!

Indeed, your pull request will receive a separate number. Please add the text `Fixes #8904` to the _body_ of your commit message on a single line (it needs to be separated from the commit _title_ by an empty line. See 4650d10ff6ad1048785a009349c8b5d6e922fc7a for an example.)

All numbers prefixed by a hash sign will be automatically converted to links by GitHub. Additionally, the "Fixes" annotation will have the effect that this issue will be automatically closed when the pull request gets merged.

**Comment by lersek** [comment]
@mspncp thank you. Could you please assign this ticket to me, meanwhile? I don't seem to have permission to do that. Thanks!

**Comment by mspncp** [comment]
> @mspncp thank you. Could you please assign this ticket to me, meanwhile? I don't seem to have permission to do that. Thanks!

Oh, sorry for not answering your request. I just noticed it now. It must have slipped through my inbox.

**Comment by mspncp** [comment]
> > @mspncp thank you. Could you please assign this ticket to me, meanwhile? I don't seem to have permission to do that. Thanks!
> 
> Oh, sorry for not answering your request. I just noticed it now. It must have slipped through my inbox.

FWIW: issues and pull requests can only applied to team members.
